sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,388 @@
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_CONFIGOPTS_REG_OFFSET, reg);
+  spi_host_enable(spi_host, true);
+  return kDifOk;
+}
+
+static void wait_ready(const dif_spi_host_t *spi_host) {
+  bool ready;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    ready = bitfield_bit32_read(reg, SPI_HOST_STATUS_READY_BIT);
+  } while (!ready);
+}
+
+static void wait_tx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t txqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+  } while (txqd == 0xFF);
+}
+
+static void wait_rx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t rxqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    rxqd = bitfield_field32_read(reg, SPI_HOST_STATUS_RXQD_FIELD);
+  } while (rxqd == 0);
+}
+
+static inline void tx_fifo_write8(const dif_spi_host_t *spi_host,
+                                  uintptr_t srcaddr) {
+  uint8_t *src = (uint8_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write8(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+static inline void tx_fifo_write32(const dif_spi_host_t *spi_host,
+                                   uintptr_t srcaddr) {
+  uint32_t *src = (uint32_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+void dif_spi_host_fifo_write(const dif_spi_host_t *spi_host, const void *src,
+                             uint16_t len) {
+  uintptr_t ptr = (uintptr_t)src;
+
+  // If the pointer starts mis-aligned, write until we are aligned.
+  while (ptr & 3 && len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+
+  // Write complete 32-bit words to the fifo.
+  while (len > 3) {
+    tx_fifo_write32(spi_host, ptr);
+    ptr += 4;
+    len -= 4;
+  }
+
+  // Clean up any leftover bytes.
+  while (len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+}
+
+typedef struct queue {
+  int32_t length;
+  union {
+    uint64_t qword;
+    uint32_t word[2];
+    uint8_t byte[8];
+  };
+} queue_t;
+
+static inline void enqueue_byte(queue_t *queue, uint8_t data) {
+  queue->byte[queue->length++] = data;
+}
+
+static inline void enqueue_word(queue_t *queue, uint32_t data) {
+  if (queue->length % sizeof(uint32_t) == 0) {
+    queue->word[queue->length / sizeof(uint32_t)] = data;
+    queue->length += 4;
+  } else {
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+  }
+}
+
+static inline uint8_t dequeue_byte(queue_t *queue) {
+  uint8_t val = queue->byte[0];
+  queue->qword >>= 8;
```

Comment:
Ah. This doesn't work. The strict aliasing rule, in genreal, forbids the following code:
```c
union { int i, float f } u = { .i = 0 };
float f = u.f;
```
since it materializes an lvalue of the wrong type and reads through it. BUT, C99 makes this *precise* pattern
legal, so long as `u` is _not_ accessed behind a pointer.

The correct way to do this would be to define the field as `uint8_t bytes[sizeof(uint64_t)]` and then use `read_32()` from memory.h (you can add a `read_64()` if you need it).

Reply:
Sigh.  Fixed.